src/wp-includes/link-template.php

```diff
@@ -417,13 +467,70 @@ function get_attachment_link( $post = null, $leavename = false ) {
 
 	$link = false;
 
-	$post   = get_post( $post );
-	$parent = ( $post->post_parent > 0 && $post->post_parent != $post->ID ) ? get_post( $post->post_parent ) : false;
-	if ( $parent && ! in_array( $parent->post_type, get_post_types(), true ) ) {
-		$parent = false;
-	}
+	$post            = get_post( $post );
+	$force_ugly_link = wp_force_ugly_post_permalink( $post );
+	$parent_id       = $post->post_parent;
+	$parent          = $parent_id ? get_post( $parent_id ) : false;
+	$parent_valid    = true; // Default for no parent.
+	if (
+		$parent_id &&
+		(
+			$post->post_parent === $post->ID ||
+			! get_post( $post->post_parent ) ||
+			! is_post_type_viewable( get_post_type( $parent_id ) )
+		)
+	) {
+		// Post is either its own parent or parent post unavailable.
+		$parent_valid = false;
+	}
+
+	if ( $force_ugly_link ) {
+		$link = false;
+	} elseif ( $wp_rewrite->using_permalinks() && ! $parent_valid ) {
```

Comment:
This condition is pretty much a copy-paste from the code in `get_permalink()` that is unreachable for attachments. For now I have left the category and filters out while working through this. 

The code is complex and messy as heck so a new private function may be the best thing to do rather than copy-paste.